common/persistence/nosql/nosqlplugin/cassandra/gocql/tracing.go

```diff
@@ -0,0 +1,141 @@
+package gocql
+
+import (
+	"context"
+	"fmt"
+	"net"
+	reflect "reflect"
+	"strings"
+	"time"
+
+	"github.com/gocql/gocql"
+	"go.opentelemetry.io/otel/attribute"
+	semconv "go.opentelemetry.io/otel/semconv/v1.12.0"
+	"go.opentelemetry.io/otel/trace"
+)
+
+const (
+	batchOperationName = "batch"
+	queryOperationName = "query"
+)
+
+var (
+	instrumentationName = reflect.TypeOf(traceObserver{}).PkgPath()
+	batchOperationAttr  = semconv.DBOperationKey.String(batchOperationName)
+	queryOperationAttr  = semconv.DBOperationKey.String(queryOperationName)
+)
+
+type traceObserver struct {
+	dbNameAttr    attribute.KeyValue
+	batchSpanName string
+	querySpanName string
+	tracer        trace.Tracer
+
+	nextBatchObserver gocql.BatchObserver
+	nextQueryObserver gocql.QueryObserver
+}
+
+// ConfigureTracing registers tracing observers with the supplied
+// gocql.ClusterConfig. Any pre-existing observers are preserved and will be
+// called after the appropriate tracing listener.
+func ConfigureTracing(config *gocql.ClusterConfig, tp trace.TracerProvider) {
+	obs := traceObserver{
+
+		// > In Cassandra, db.name SHOULD be set to the keyspace name
+		// https://github.com/open-telemetry/opentelemetry-specification/blob/4f8d9f679d8e138693eb6fd085e6f6654012d116/specification/trace/semantic_conventions/database.md?plain=1#L179
+		dbNameAttr: semconv.DBNameKey.String(config.Keyspace),
+
+		// > If db.sql.table is not available [...] the span SHOULD be named
+		// > "<db.operation> <db.name>"
+		// https://github.com/open-telemetry/opentelemetry-specification/blob/4f8d9f679d8e138693eb6fd085e6f6654012d116/specification/trace/semantic_conventions/database.md?plain=1#L29
+		batchSpanName:     fmt.Sprintf("%s %s", batchOperationName, config.Keyspace),
+		querySpanName:     fmt.Sprintf("%s %s", queryOperationName, config.Keyspace),
+		tracer:            tp.Tracer(instrumentationName),
+		nextBatchObserver: config.BatchObserver,
+		nextQueryObserver: config.QueryObserver,
+	}
+	config.BatchObserver = &obs
+	config.QueryObserver = &obs
+}
+
+// ObserveBatch handles callbacks from gocql upon completion of a Cassandra
+// batch operation.
+func (sto *traceObserver) ObserveBatch(
```

Comment:
why `sto`?

Reply:
Oh good catch. The type was sessionTraceObserver for a while. I'll fix.